JUnit multiple levels of Nested fail with QuarkusTest

### Describe the bug

If there is more than one level of `@NestedTests` in a JUnit test with `@QuarkusTest`, each one throws an `org.junit.jupiter.api.extension.TestInstantiationException` due to a `javax.enterprise.inject.UnsatisfiedResolutionException`.

However, the same setup without QuarkusTest works.

*Example:*
```
@QuarkusTest
public class MultipleNestedQuarkusTest {

    @Nested
    class FirstNested {

        @Nested
        class SecondNested {

            @Test
            void check() {
                Assertions.assertTrue(true);
            }
        }
    }
}
```

### Expected behavior

Above test passes.

### Actual behavior

Stacktrace when executing test above.

```
org.junit.jupiter.api.extension.TestInstantiationException: Failed to create test instance

	at io.quarkus.test.junit.QuarkusTestExtension.initTestState(QuarkusTestExtension.java:848)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:812)

...

Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.runner.bootstrap.RunningQuarkusApplicationImpl.instance(RunningQuarkusApplicationImpl.java:83)
	... 97 more
Caused by: javax.enterprise.inject.UnsatisfiedResolutionException: No bean found for required type [class org.acme.MultipleNestedQuarkusTest$FirstNested] and qualifiers [[]]
	at io.quarkus.arc.impl.InstanceImpl.bean(InstanceImpl.java:175)
	at io.quarkus.arc.impl.InstanceImpl.getInternal(InstanceImpl.java:196)
	at io.quarkus.arc.impl.InstanceImpl.get(InstanceImpl.java:93)
	... 102 more
```

### How to Reproduce?

Run the test above.

Reproducer: https://github.com/vonatzigenc/reproducer-nested-quarkustest/blob/main/src/test/java/org/acme/MultipleNestedQuarkusTest.java

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

openjdk version "11.0.12" 2021-07-20 LTS

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.2.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.1

### Additional information

_No response_

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/QuarkusTestNestedTestCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/QuarkusTestNestedTestCase.java
@@ -13,6 +13,7 @@ import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Order;
 import org.junit.jupiter.api.Tag;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
 import org.junit.jupiter.api.TestMethodOrder;
 
 import io.quarkus.test.junit.QuarkusTest;
@@ -35,6 +36,8 @@ public class QuarkusTestNestedTestCase {
     private static final AtomicInteger COUNT_AFTER_ALL = new AtomicInteger(0);
     private static final String EXPECTED_OUTER_VALUE = "set from outer";
     private static final String EXPECTED_INNER_VALUE = "set from inner";
+    private static final String EXPECTED_SECOND_LEVEL_FIRST_INNER_VALUE = "set from second level first inner";
+    private static final String EXPECTED_SECOND_LEVEL_SECOND_INNER_VALUE = "set from second level second inner";
 
     String outerValue;
 
@@ -74,9 +77,9 @@ public class QuarkusTestNestedTestCase {
         @Order(1)
         void testOne() {
             assertEquals(1, COUNT_BEFORE_ALL.get(), "COUNT_BEFORE_ALL");
-            assertEquals(5, COUNT_BEFORE_EACH.get(), "COUNT_BEFORE_EACH");
+            assertEquals(7, COUNT_BEFORE_EACH.get(), "COUNT_BEFORE_EACH");
             assertEquals(2, COUNT_TEST.getAndIncrement(), "COUNT_TEST");
-            assertEquals(3, COUNT_AFTER_EACH.get(), "COUNT_AFTER_EACH");
+            assertEquals(5, COUNT_AFTER_EACH.get(), "COUNT_AFTER_EACH");
             assertEquals(0, COUNT_AFTER_ALL.get(), "COUNT_AFTER_ALL");
         }
 
@@ -84,9 +87,9 @@ public class QuarkusTestNestedTestCase {
         @Order(2)
         void testTwo() {
             assertEquals(1, COUNT_BEFORE_ALL.get(), "COUNT_BEFORE_ALL");
-            assertEquals(7, COUNT_BEFORE_EACH.get(), "COUNT_BEFORE_EACH");
+            assertEquals(9, COUNT_BEFORE_EACH.get(), "COUNT_BEFORE_EACH");
             assertEquals(3, COUNT_TEST.getAndIncrement(), "COUNT_TEST");
-            assertEquals(5, COUNT_AFTER_EACH.get(), "COUNT_AFTER_EACH");
+            assertEquals(7, COUNT_AFTER_EACH.get(), "COUNT_AFTER_EACH");
             assertEquals(0, COUNT_AFTER_ALL.get(), "COUNT_AFTER_ALL");
         }
 
@@ -99,6 +102,38 @@ public class QuarkusTestNestedTestCase {
         @AfterEach
         void afterEach() {
             COUNT_AFTER_EACH.incrementAndGet();
+        }
+
+        @Nested
+        @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+        @TestMethodOrder(OrderAnnotation.class)
+        class SecondLevelInnerNested {
+
+            private final AtomicInteger SECOND_LEVEL_COUNTER = new AtomicInteger(0);
+
+            String secondLevelInnerValue;
+
+            @BeforeEach
+            void beforeEach() {
+                SECOND_LEVEL_COUNTER.incrementAndGet();
+                secondLevelInnerValue = EXPECTED_SECOND_LEVEL_FIRST_INNER_VALUE;
+            }
+
+            @Test
+            @Order(1)
+            void testOne() {
+                // assertEquals(1, SECOND_LEVEL_COUNTER.get(), "SECOND_LEVEL_COUNTER");
+                assertEquals(1, SECOND_LEVEL_COUNTER.get(), "SECOND_LEVEL_COUNTER");
+            }
+
+            @Test
+            @Order(2)
+            void testSecondLevelAndInnerAndOuterValues() {
+                assertEquals(2, SECOND_LEVEL_COUNTER.get(), "SECOND_LEVEL_COUNTER");
+                assertEquals(EXPECTED_INNER_VALUE, innerValue);
+                assertEquals(EXPECTED_OUTER_VALUE, outerValue);
+                assertEquals(EXPECTED_SECOND_LEVEL_FIRST_INNER_VALUE, secondLevelInnerValue);
+            }
         }
     }
 
@@ -123,6 +158,23 @@ public class QuarkusTestNestedTestCase {
         void afterEach() {
             COUNT_AFTER_EACH.incrementAndGet();
         }
+
+        @Nested
+        class SecondLevelInnerNested {
+
+            String secondLevelInnerValue;
+
+            @BeforeEach
+            void beforeEach() {
+                secondLevelInnerValue = EXPECTED_SECOND_LEVEL_SECOND_INNER_VALUE;
+            }
+
+            @Test
+            void testSecondLevelAndInnerAndOuterValues() {
+                assertEquals(EXPECTED_OUTER_VALUE, outerValue);
+                assertEquals(EXPECTED_SECOND_LEVEL_SECOND_INNER_VALUE, secondLevelInnerValue);
+            }
+        }
     }
 
     @AfterEach
@@ -133,9 +185,9 @@ public class QuarkusTestNestedTestCase {
     @AfterAll
     static void afterAll() {
         assertEquals(1, COUNT_BEFORE_ALL.get(), "COUNT_BEFORE_ALL");
-        assertEquals(9, COUNT_BEFORE_EACH.get(), "COUNT_BEFORE_EACH");
+        assertEquals(15, COUNT_BEFORE_EACH.get(), "COUNT_BEFORE_EACH");
         assertEquals(4, COUNT_TEST.get(), "COUNT_TEST");
-        assertEquals(9, COUNT_AFTER_EACH.get(), "COUNT_AFTER_EACH");
+        assertEquals(15, COUNT_AFTER_EACH.get(), "COUNT_AFTER_EACH");
         assertEquals(0, COUNT_AFTER_ALL.getAndIncrement(), "COUNT_AFTER_ALL");
     }
 }

--- a/test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/internal/ResetOuterMockitoMocksCallback.java
+++ b/test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/internal/ResetOuterMockitoMocksCallback.java
@@ -7,8 +7,10 @@ public class ResetOuterMockitoMocksCallback implements QuarkusTestAfterAllCallba
 
     @Override
     public void afterAll(QuarkusTestContext context) {
-        if (context.getOuterInstance() != null) {
-            MockitoMocksTracker.reset(context.getOuterInstance());
+        if (context.getOuterInstances() != null) {
+            for (Object outerInstance : context.getOuterInstances()) {
+                MockitoMocksTracker.reset(outerInstance);
+            }
         }
     }
 }

--- a/test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/internal/SetMockitoMockAsBeanMockCallback.java
+++ b/test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/internal/SetMockitoMockAsBeanMockCallback.java
@@ -9,8 +9,10 @@ public class SetMockitoMockAsBeanMockCallback implements QuarkusTestBeforeEachCa
     @Override
     public void beforeEach(QuarkusTestMethodContext context) {
         MockitoMocksTracker.getMocks(context.getTestInstance()).forEach(this::installMock);
-        if (context.getOuterInstance() != null) {
-            MockitoMocksTracker.getMocks(context.getOuterInstance()).forEach(this::installMock);
+        if (context.getOuterInstances() != null) {
+            for (Object outerInstance : context.getOuterInstances()) {
+                MockitoMocksTracker.getMocks(outerInstance).forEach(this::installMock);
+            }
         }
     }
 

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
@@ -17,6 +17,7 @@ import java.time.temporal.ChronoUnit;
 import java.util.AbstractMap;
 import java.util.ArrayList;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
@@ -120,7 +121,7 @@ public class QuarkusTestExtension extends AbstractJvmQuarkusTestExtension
     private static Class<?> actualTestClass;
     private static Object actualTestInstance;
     // needed for @Nested
-    private static Object outerInstance;
+    private static List<Object> outerInstances = new ArrayList<>(1);
     private static RunningQuarkusApplication runningQuarkusApplication;
     private static Pattern clonePattern;
     private static Throwable firstException; //if this is set then it will be thrown from the very first test that is run, the rest are aborted
@@ -558,9 +559,9 @@ public class QuarkusTestExtension extends AbstractJvmQuarkusTestExtension
             throw new RuntimeException("Could not find method " + originalTestMethod + " on test class");
         }
 
-        Constructor<?> constructor = quarkusTestMethodContextClass.getConstructor(Object.class, Object.class, Method.class);
+        Constructor<?> constructor = quarkusTestMethodContextClass.getConstructor(Object.class, List.class, Method.class);
         return new AbstractMap.SimpleEntry<>(quarkusTestMethodContextClass,
-                constructor.newInstance(actualTestInstance, outerInstance, actualTestMethod));
+                constructor.newInstance(actualTestInstance, outerInstances, actualTestMethod));
     }
 
     private boolean isNativeOrIntegrationTest(Class<?> clazz) {
@@ -760,13 +761,18 @@ public class QuarkusTestExtension extends AbstractJvmQuarkusTestExtension
         try {
             actualTestClass = Class.forName(extensionContext.getRequiredTestClass().getName(), true,
                     Thread.currentThread().getContextClassLoader());
-            outerInstance = null;
             if (extensionContext.getRequiredTestClass().isAnnotationPresent(Nested.class)) {
                 Class<?> outerClass = actualTestClass.getEnclosingClass();
-                outerInstance = runningQuarkusApplication.instance(outerClass);
                 Constructor<?> declaredConstructor = actualTestClass.getDeclaredConstructor(outerClass);
                 declaredConstructor.setAccessible(true);
-                actualTestInstance = declaredConstructor.newInstance(outerInstance);
+                if (outerClass.getEnclosingClass() != null) {
+                    outerInstances.add(actualTestInstance);
+                    actualTestInstance = declaredConstructor.newInstance(actualTestInstance);
+                } else {
+                    Object outerInstance = runningQuarkusApplication.instance(outerClass);
+                    actualTestInstance = declaredConstructor.newInstance(outerInstance);
+                    outerInstances.add(outerInstance);
+                }
             } else {
                 actualTestInstance = runningQuarkusApplication.instance(actualTestClass);
             }
@@ -780,7 +786,7 @@ public class QuarkusTestExtension extends AbstractJvmQuarkusTestExtension
                 afterConstructCallback.getClass().getMethod("afterConstruct", Object.class).invoke(afterConstructCallback,
                         actualTestInstance);
             }
-            if (outerInstance != null) {
+            for (Object outerInstance : outerInstances) {
                 for (Object afterConstructCallback : afterConstructCallbacks) {
                     afterConstructCallback.getClass().getMethod("afterConstruct", Object.class).invoke(afterConstructCallback,
                             outerInstance);
@@ -905,14 +911,20 @@ public class QuarkusTestExtension extends AbstractJvmQuarkusTestExtension
         try {
             Class<?> testClassFromTCCL = Class.forName(extensionContext.getRequiredTestClass().getName(), true,
                     Thread.currentThread().getContextClassLoader());
-            Object effectiveTestInstance = actualTestInstance;
-            Method newMethod = determineTCCLExtensionMethod(invocationContext.getExecutable(), testClassFromTCCL);
+            Map<Class<?>, Object> allTestsClasses = new HashMap<>();
+            // static loading
+            allTestsClasses.put(testClassFromTCCL, actualTestInstance);
             // this is needed to support before*** and after*** methods that are part of class that encloses the test class
             // (the test class is in this case a @Nested test)
-            if ((newMethod == null) && (testClassFromTCCL.getEnclosingClass() != null)) {
-                newMethod = determineTCCLExtensionMethod(invocationContext.getExecutable(),
-                        testClassFromTCCL.getEnclosingClass());
-                effectiveTestInstance = outerInstance;
+            outerInstances.forEach(i -> allTestsClasses.put(i.getClass(), i));
+            Method newMethod = null;
+            Object effectiveTestInstance = null;
+            for (Map.Entry<Class<?>, Object> testClass : allTestsClasses.entrySet()) {
+                newMethod = determineTCCLExtensionMethod(invocationContext.getExecutable(), testClass.getKey());
+                if (newMethod != null) {
+                    effectiveTestInstance = testClass.getValue();
+                    break;
+                }
             }
 
             if (newMethod == null) {
@@ -1066,7 +1078,7 @@ public class QuarkusTestExtension extends AbstractJvmQuarkusTestExtension
             }
         } finally {
             currentTestClassStack.pop();
-            outerInstance = null;
+            outerInstances.clear();
         }
     }
 
@@ -1080,8 +1092,8 @@ public class QuarkusTestExtension extends AbstractJvmQuarkusTestExtension
 
         Class<?> quarkusTestContextClass = Class.forName(QuarkusTestContext.class.getName(), true,
                 runningQuarkusApplication.getClassLoader());
-        Object quarkusTestContextInstance = quarkusTestContextClass.getConstructor(Object.class, Object.class)
-                .newInstance(actualTestInstance, outerInstance);
+        Object quarkusTestContextInstance = quarkusTestContextClass.getConstructor(Object.class, List.class)
+                .newInstance(actualTestInstance, outerInstances);
 
         ClassLoader original = setCCL(runningQuarkusApplication.getClassLoader());
         try {

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/callback/QuarkusTestContext.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/callback/QuarkusTestContext.java
@@ -1,23 +1,25 @@
 package io.quarkus.test.junit.callback;
 
+import java.util.List;
+
 /**
  * Context object passed to {@link QuarkusTestAfterAllCallback}
  */
 public class QuarkusTestContext {
 
     private final Object testInstance;
-    private final Object outerInstance;
+    private final List<Object> outerInstances;
 
-    public QuarkusTestContext(Object testInstance, Object outerInstance) {
+    public QuarkusTestContext(Object testInstance, List<Object> outerInstances) {
         this.testInstance = testInstance;
-        this.outerInstance = outerInstance;
+        this.outerInstances = outerInstances;
     }
 
     public Object getTestInstance() {
         return testInstance;
     }
 
-    public Object getOuterInstance() {
-        return outerInstance;
+    public List<Object> getOuterInstances() {
+        return outerInstances;
     }
 }

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/callback/QuarkusTestMethodContext.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/callback/QuarkusTestMethodContext.java
@@ -1,6 +1,7 @@
 package io.quarkus.test.junit.callback;
 
 import java.lang.reflect.Method;
+import java.util.List;
 
 /**
  * Context object passed to {@link QuarkusTestBeforeEachCallback} and {@link QuarkusTestAfterEachCallback}
@@ -9,8 +10,8 @@ public final class QuarkusTestMethodContext extends QuarkusTestContext {
 
     private final Method testMethod;
 
-    public QuarkusTestMethodContext(Object testInstance, Object outerInstance, Method testMethod) {
-        super(testInstance, outerInstance);
+    public QuarkusTestMethodContext(Object testInstance, List<Object> outerInstances, Method testMethod) {
+        super(testInstance, outerInstances);
         this.testMethod = testMethod;
     }
 

